includes/class-wc-customer-download-log.php

```diff
@@ -0,0 +1,173 @@
+
+		if ( is_numeric( $download_log ) && $download_log > 0 ) {
+			$this->set_id( $download_log );
+		} elseif ( $download_log instanceof self ) {
+			$this->set_id( $download_log->get_id() );
+		} elseif ( is_object( $download_log ) && ! empty( $download_log->download_log_id ) ) {
+			$this->set_id( $download_log->download_log_id );
+			$this->set_props( (array) $download_log );
+			$this->set_object_read( true );
+		} else {
+			$this->set_object_read( true );
+		}
+
+		$this->data_store = WC_Data_Store::load( 'customer-download-log' );
+
+		if ( $this->get_id() > 0 ) {
+			$this->data_store->read( $this );
+		}
+ 	}
+
+	/*
+	|--------------------------------------------------------------------------
+	| Getters
+	|--------------------------------------------------------------------------
+	*/
+
+	/**
+	 * Get timestamp.
+	 *
+	 * @param  string $context
+	 * @return WC_DateTime|null Object if the date is set or null if there is no date.
+	 */
+	public function get_timestamp( $context = 'view' ) {
+		return $this->get_prop( 'timestamp', $context );
+	}
+
+	/**
+	 * Get permission id.
+	 *
+	 * @param  string $context
+	 * @return integer
+	 */
+	public function get_permission_id( $context = 'view' ) {
+		return $this->get_prop( 'permission_id', $context );
+	}
+
+	/**
+	 * Get user id.
+	 *
+	 * @param  string $context
+	 * @return integer
+	 */
+	public function get_user_id( $context = 'view' ) {
+		return $this->get_prop( 'user_id', $context );
+	}
+
+	/**
+	 * Get user ip address.
+	 *
+	 * @param  string $context
+	 * @return string
+	 */
+	public function get_user_ip_address( $context = 'view' ) {
+		return $this->get_prop( 'user_ip_address', $context );
+	}
+
+	/*
+	|--------------------------------------------------------------------------
+	| Setters
+	|--------------------------------------------------------------------------
+	*/
+
+	/**
+	 * Set timestamp.
+	 *
+	 * @param string|integer|null $date UTC timestamp, or ISO 8601 DateTime. If the DateTime string has no timezone or offset, WordPress site timezone will be assumed. Null if their is no date.
+	 */
+	public function set_timestamp( $date = null ) {
+		$this->set_date_prop( 'timestamp', $date );
+	}
+
+	/**
+	 * Set permission id.
+	 *
+	 * @param int $value
+	 */
+	public function set_permission_id( $value ) {
+		$this->set_prop( 'permission_id', absint( $value ) );
+	}
+
+	/**
+	 * Set user id.
+	 *
+	 * @param int $value
+	 */
+	public function set_user_id( $value ) {
+		$this->set_prop( 'user_id', absint( $value ) );
+	}
+
+	/**
+	 * Set user ip address.
+	 *
+	 * @param string $value
+	 */
+	public function set_user_ip_address( $value ) {
+		$this->set_prop( 'user_ip_address', $value );
+	}
+
+	/*
+	|--------------------------------------------------------------------------
+	| CRUD methods
+	|--------------------------------------------------------------------------
+	*/
+
+	/**
+	 * Save data to the database.
+	 *
+	 * @return int Log ID
+	 */
+	public function save() {
```

Comment:
This method is exactly the same as `WC_Data::save`. You can delete it and just let the parent method do the work.